spec_validator.go

```diff
@@ -0,0 +1,148 @@
+/**
+ *  Copyright (c) 2022  Comcast Cable Communications Management, LLC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+
+package wrp
+
+import (
+	"errors"
+	"fmt"
+	"regexp"
+	"strconv"
+	"strings"
+	"unicode"
+
+	"github.com/google/uuid"
+)
+
+const (
+	serialPrefix = "serial"
+	uuidPrefix   = "uuid"
+	eventPrefix  = "event"
+	dnsPrefix    = "dns"
+)
+
+var (
+	ErrorInvalidMessageEncoding = errors.New("invalid message encoding")
+	ErrorInvalidMessageType     = errors.New("invalid message type")
+	ErrorInvalidSource          = errors.New("invalid Source name")
+	ErrorInvalidDestination     = errors.New("invalid Destination name")
+)
+
+// SpecValidators is a WRP validator that ensures messages are valid based on
+// each spec validator in the list. Only validates the opinionated portions of the spec.
+// SpecValidators validates the following fields: UTF8 (all string fields), MessageType, Source, Destination
+func SpecValidators() Validators {
+	return Validators{UTF8Validator, MessageTypeValidator, SourceValidator, DestinationValidator}
+}
+
+// UTF8Validator is a WRP validator that takes messages and validates that it contains UTF-8 strings.
+var UTF8Validator ValidatorFunc = func(m Message) error {
+	if err := UTF8(m); err != nil {
+		return fmt.Errorf("%w: %v", ErrorInvalidMessageEncoding, err)
+	}
+
+	return nil
+}
+
+// MessageTypeValidator is a WRP validator that takes messages and validates their Type.
+var MessageTypeValidator ValidatorFunc = func(m Message) error {
+	t := m.MessageType()
+	if t < Invalid0MessageType || t > lastMessageType {
+		return ErrorInvalidMessageType
+	}
+
+	switch t {
+	case Invalid0MessageType, Invalid1MessageType, lastMessageType:
+		return ErrorInvalidMessageType
+	}
+
+	return nil
+}
+
+// SourceValidator is a WRP validator that takes messages and validates their Source.
+// Only mac and uuid sources are validated. Serial, event and dns sources are
+// not validated.
+var SourceValidator ValidatorFunc = func(m Message) error {
+	if err := validateLocator(m.Source); err != nil {
+		return fmt.Errorf("%w: %v", ErrorInvalidSource, err)
```

Comment:
I don't think the source is sensitive information so we might as well include them in the error for any human who reads it. 🙂  The same can be done when validating the destination.
```suggestion
		return fmt.Errorf("%w '%s': %v", ErrorInvalidSource, m.Source, err)
```